internal/grpc/internalerrs/prometheus.go

```
@@ -0,0 +1,107 @@
+},
+	[]string{
+		"grpc_service",      // e.g. "gitserver.v1.GitserverService"
+		"grpc_method",       // e.g. "Exec"
+		"grpc_code",         // e.g. "NotFound"
+		"is_internal_error", // e.g. "true"
+	},
+)
+
+// PrometheusUnaryClientInterceptor returns a grpc.UnaryClientInterceptor that observes the result of
+// the RPC and records it as a Prometheus metric ("src_grpc_method_status").
+func PrometheusUnaryClientInterceptor(ctx context.Context, fullMethod string, req, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	serviceName, methodName := splitMethodName(fullMethod)
+
+	err := invoker(ctx, fullMethod, req, reply, cc, opts...)
+	doObservation(serviceName, methodName, err)
+	return err
+}
+
+// PrometheusStreamClientInterceptor returns a grpc.StreamClientInterceptor that observes the result of
+// the RPC and records it as a Prometheus metric ("src_grpc_method_status").
+//
+// If any errors are encountered during the stream, the first error is recorded. Otherwise, the
+// final status of the stream is recorded.
+func PrometheusStreamClientInterceptor(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, fullMethod string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	serviceName, methodName := splitMethodName(fullMethod)
+
+	s, err := streamer(ctx, desc, cc, fullMethod, opts...)
+	if err != nil {
+		doObservation(serviceName, methodName, err) // method failed to be invoked at all, record it
+		return nil, err
+	}
+
+	return newPrometheusServerStream(s, serviceName, methodName), err
+}
+
+// newPrometheusServerStream wraps a grpc.ClientStream to observe the first error
+// encountered during the stream, if any.
+func newPrometheusServerStream(s grpc.ClientStream, serviceName, methodName string) grpc.ClientStream {
+	var observeOnce sync.Once
```

It's not super clear to me that sending and receiving should share the same `observeOnce`. Sending can return an error independently of (and in addition to) an error from receive. This is probably an okay simplification, but I wanted to at least mention it in case you hadn't already thought it through

---

I addressed this in [`3e750c1` (#51749)](https://github.com/sourcegraph/sourcegraph/pull/51749/commits/3e750c19c799dcf6bc5e9e494a61afe6ada878a1), but I'm copying it here too:

> 	// Design note: We only want a single observation for each RPC call: it either succeeds or fails
	// with a single error. This ensures we do not double-count RPCs in Prometheus metrics.
	//
	// For unary calls this is straightforward, but for streaming RPCs we need to make a compromise. We only
	// observe the first error (either sending or receiving) that occurs during the stream, instead of every
	// error that occurs during the stream's lifespan. While this approach swallows some errors, it keeps the
	// Prometheus metric count clean and non-duplicated. The logging interceptor handles surfacing all errors
	// that are encountered during a stream.